webui/src/App.js

```diff
@@ -21,86 +21,168 @@ import {
+      clearTimeout(timeoutId);
+      fetch('/defaults').then(res => res.json()).then(data => {
+        if (!cleaningUp) {
+          setDefaults(data);
+        }
+      }).catch(reason => {
+        if (!cleaningUp) {
+          timeoutId = setTimeout(getDefaults, 1000);
+        }
+      });
     }
-  });
 
-  ws.addEventListener('error', function(ev) {
-    ws.close();
-  });
+    if (!defaults) {
+      getDefaults();
+    }
+
+    return () => {
+      cleaningUp = true;
+      clearTimeout(timeoutId);
+    };
+  }, [defaults]);
+
+  return { defaults, reloadDefaults };
+}
 
-  ws.addEventListener('close', function(ev) {
-    setTimeout(connect, 1000);
+// returns { curValuesRef, emit, isWsReady, wsCallbacksRef }
+// curValuesRef tracks values as well as thresholds.
+// Use emit to send events to the server.
+// isWsReady indicates that the websocket is connected and has received
+// its first values reading from the server.
+// onCloseWs is called when the websocket closes for any reason,
+// unless the hooks is already doing effect cleanup.
+// The expectation is that onCloseWs is used to reload default thresholds
+// and provide new ones to trigger a new connection.
+function useWsConnection({ defaults, onCloseWs }) {
+  const [isWsReady, setIsWsReady] = useState(false);
+
+  const curValuesRef = useRef({
+
+    // A history of the past 'max_size' values fetched from the backend.
+    // Used for plotting and displaying live values.
+    // We use a cyclical array to save memory.
+    kCurValues: [],
+    oldest: 0,
+
+    // Keep track of the current thresholds fetched from the backend.
+    kCurThresholds: [],
   });
 
-  ws.addEventListener('message', function(ev) {
-    const data = JSON.parse(ev.data)
-    const action = data[0];
-    const msg = data[1];
+  const wsRef = useRef();
+  const wsCallbacksRef = useRef({});
 
-    if (wsCallbacks[action]) {
-      wsCallbacks[action](msg);
+  const emit = useCallback((msg) => {
+    // App should wait for isWsReady to send messages.
+    if (!wsRef.current || !isWsReady) {
+      throw new Error("emit() called when isWsReady !== true.");
     }
-  });
-}
 
-function emit(msg) {
-  // Queue the message if the websocket connection is not ready yet.
-  // The states are CONNECTING (0), OPEN (1), CLOSING (2) and CLOSED (3).
-  if (ws.readyState !== 1 /* OPEN */) {
-    wsQueue.push(ws);
-    return;
-  }
+    wsRef.current.send(JSON.stringify(msg));
+  }, [isWsReady, wsRef]);
 
-  ws.send(JSON.stringify(msg));
-}
+  wsCallbacksRef.current.values = function(msg) {
+    const curValues = curValuesRef.current;
+    if (curValues.kCurValues.length < max_size) {
+      curValues.kCurValues.push(msg.values);
+    } else {
+      curValues.kCurValues[curValues.oldest] = msg.values;
+      curValues.oldest = (curValues.oldest + 1) % max_size;
+    }
+    if (curValues.kCurValues.length === 1) {
+      // WebSocket is considered ready after it gets its first "values" message.
```

Comment:
I decided to wait for the server to send a sensor reading before loading the rest of the UI, instead of defaulting to 0's. It would be straightforward to switch back the "initialize to 0's" behavior too.

Reply:
I think switching back to to the previous behavior is better only really for one reason.

I'm working on adding a config page with a dropdown that lets you select the serial device to connect to from the UI. Because of this, the UI would need to be loaded before the values message has potentially been sent.

Reply:
That's a great reason, UI selectable serial ports sound very useful.